### PR TITLE
#2677 - Impacted application reassessment on COE declined

### PR DESCRIPTION
**Pull request by dheepak-aot**
## Assess Impacted applications to create reassessment when COE is declined

- [x] When institution decline a COE (Either first or second COE), call the centralized logic to assess for impacted application and create reassessment if applicable.

**Application 1:**

image

**Application 2:** 

image

**Institution Decline COE for Application 1:**

image

**Related application reassessment for Application 2:**

image

- [x] If cancellation happens for an application for which the COE is declined, then do NOT assess for impacted application for such applications as the `COE Decline` action will handle this part.

**Then application 1 is cancelled:**

image

**The application 2 is NOT re-assessed again:**

image

- [x] Create E2E tests for `Institution Decline COE` and `Student cancel an application with Declined COE`

image

image

### CODE REVIEW

File: sources/packages/backend/apps/api/src/route-controllers/confirmation-of-enrollment/_tests_/e2e/confirmation-of-enrollment.institutions.denyConfirmationOfEnrollment.e2e-spec.ts [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+  AssessmentTriggerType,
+  COEStatus,
+  Institution,
+  InstitutionLocation,
+  OfferingIntensity,
+  StudentAssessmentStatus,
+} from "@sims/sims-db";
+import { addDays } from "@sims/utilities";
+
+describe("ConfirmationOfEnrollmentInstitutionsController(e2e)-confirmEnrollment", () => {
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/route-controllers/confirmation-of-enrollment/_tests_/e2e/confirmation-of-enrollment.institutions.denyConfirmationOfEnrollment.e2e-spec.ts:30]: The title should be adjusted as below.
```
ConfirmationOfEnrollmentInstitutionsController(e2e)-denyConfirmationOfEnrollment
```

File: sources/packages/backend/apps/api/src/route-controllers/confirmation-of-enrollment/_tests_/e2e/confirmation-of-enrollment.institutions.denyConfirmationOfEnrollment.e2e-spec.ts [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+          student,
+          institution: collegeC,
+          institutionLocation: collegeCLocation,
+        },
+        {
+          offeringIntensity: OfferingIntensity.partTime,
+          applicationStatus: ApplicationStatus.Completed,
+          currentAssessmentInitialValues: {
+            assessmentWorkflowId: "some fake id",
+            assessmentDate: addDays(-1, new Date()),
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/route-controllers/confirmation-of-enrollment/_tests_/e2e/confirmation-of-enrollment.institutions.denyConfirmationOfEnrollment.e2e-spec.ts:109]: I came to realize later that `addDays(-1, new Date())` can be simplified as `addDays(-1)`.

File: sources/packages/backend/apps/api/src/route-controllers/confirmation-of-enrollment/_tests_/e2e/confirmation-of-enrollment.institutions.denyConfirmationOfEnrollment.e2e-spec.ts [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+
+      const [firstDisbursementSchedule] =
+        applicationToDeclineCOE.currentAssessment.disbursementSchedules;
+      const endpoint = `/institutions/location/${collegeCLocation.id}/confirmation-of-enrollment/disbursement-schedule/${firstDisbursementSchedule.id}/deny`;
+      const coeDenyReasonId = 2;
+      // Act/Assert
+      await request(app.getHttpServer())
+        .patch(endpoint)
+        .send({ coeDenyReasonId })
+        .auth(
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/route-controllers/confirmation-of-enrollment/_tests_/e2e/confirmation-of-enrollment.institutions.denyConfirmationOfEnrollment.e2e-spec.ts:161]: The token would be usually acquired at the `arrange`.

File: sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts [hunk 4]
```diff
@@ -32,6 +33,9 @@ import {
   StudentAssessmentStatus,
 } from "@sims/sims-db";
 import * as faker from "faker";
+import { AssessmentSequentialProcessingService } from "@sims/services";
+import { TestingModule } from "@nestjs/testing";
+import { QueueConsumersModule } from "../../../../src/queue-consumers.module";
 
 jest.setTimeout(1200000);
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts:40]: Not part of this PR, but would you mind removing this `jest.setTimeout(1200000);`?

File: sources/packages/backend/apps/queue-consumers/src/processors/assessment/cancel-application-assessment.processor.ts [hunk 5]
```diff
@@ -145,11 +146,19 @@ export class CancelApplicationAssessmentProcessor {
         transactionEntityManager,
       );
       await summary.info("Overawards rollback check executed.");
+      // Check if the assessment has no COE(s) which were declined.
+      const hasNoDeclinedCOE = assessment.disbursementSchedules.every(
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/assessment/cancel-application-assessment.processor.ts:150]: Completly minor for this scenario, but my first instinct would be to do the `some` as below. Is this way it would "stop" at the first currency instead of checking all. It is minor because it will be 1 or 2 items 😄 
```ts
assessment.disbursementSchedules.some((disbursement) => disbursement.coeStatus === COEStatus.declined)
```

File: sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts [hunk 6]
```diff
@@ -489,6 +495,97 @@ describe(
...
+              currentAssessmentInitialValues: {
+                assessmentWorkflowId: "some fake id",
+                assessmentDate: addDays(1, new Date()),
+                studentAssessmentStatus: StudentAssessmentStatus.Submitted,
+              },
+            },
+          );
+        // Create related application reassessment for the future application.
+        // Replicating the effect of COE getting declined for current application.
+        const impactedApplicationAssessmentBeforeCancel =
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts:539]: If my understanding is correct the assert can be executed even without creating this assessment.
The creation of this record is to bring the scenario being tested closer to a real-world scenario, right?

File: sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts [hunk 7]
```diff
@@ -489,6 +495,97 @@ describe(
...
+        const assessmentSequentialProcessingService =
+          await getProviderInstanceForModule(
+            testingModule,
+            QueueConsumersModule,
+            AssessmentSequentialProcessingService,
+          );
+        const sequentialProcessingServiceSpy = jest.spyOn(
+          assessmentSequentialProcessingService,
+          "assessImpactedApplicationReassessmentNeeded",
+        );
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts:584]: Does it not have to be created before the `Act` method is called?
I assume that this will always be true.

File: sources/packages/backend/apps/queue-consumers/src/processors/assessment/cancel-application-assessment.processor.ts [hunk 8]
```diff
@@ -145,11 +146,19 @@ export class CancelApplicationAssessmentProcessor {
...
+        (disbursement) => disbursement.coeStatus !== COEStatus.declined,
+      );
       if (
         assessment.application.applicationStatus !==
-        ApplicationStatus.Overwritten
+          ApplicationStatus.Overwritten &&
+        hasNoDeclinedCOE
       ) {
         // Overwritten applications do not cause impacts in future application and the checks can be completely skipped.
+        // If a COE of an application is declined, during the process of COE being declined
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/assessment/cancel-application-assessment.processor.ts:159]: I suggest minor adjustments to the comment for improved readability.
```
If a COE of an application is declined, during the process of COE being declined, the impacted application if exists is identified and reassessed. Hence while cancelling an application with a declined COE, it must NOT be assessed for impacted application reassessment again.
```

File: sources/packages/backend/libs/services/src/confirmation-of-enrollment/confirmation-of-enrollment.service.ts [hunk 9]
```diff
@@ -627,7 +655,7 @@ export class ConfirmationOfEnrollmentService {
       );
     }
 
-    await this.updateCOEToDeclined(disbursementSchedule.id, auditUserId, {
+    await this.updateCOEToDeclined(disbursementSchedule, auditUserId, {
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/services/src/confirmation-of-enrollment/confirmation-of-enrollment.service.ts:658]: Can we change this to be explicit about the parameters being expected?
Instead of providing the full `disbursementSchedule` we should probably provide the `disbursementScheduleId` and the `studentAssessmentId`.

**Review by andrewsignori-aot — COMMENTED** [review]
Great work, please take a look at the comments.

File: sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts (hunk 6 above)
Comment (dheepak-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts:539]: Yes, exactly for being tested for close to real world scenario the assessment was added.

File: sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts (hunk 7 above)
Comment (dheepak-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts:584]: You are right. Nice catch.

File: sources/packages/backend/libs/services/src/confirmation-of-enrollment/confirmation-of-enrollment.service.ts (hunk 9 above)
Comment (dheepak-aot) [reply sources/packages/backend/libs/services/src/confirmation-of-enrollment/confirmation-of-enrollment.service.ts:658]: I can update that way, as potentially this method is expected to go through a bug fix to decline second enrolment automatically, when first is declined I thought it is going to receive more properties from disbursement object.

File: sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts [hunk 10]
```diff
@@ -489,6 +493,98 @@ describe(
...
+        const futureImpactedApplication =
+          await saveFakeApplicationDisbursements(
+            db.dataSource,
+            { student },
+            {
+              offeringIntensity: OfferingIntensity.partTime,
+              applicationStatus: ApplicationStatus.Completed,
+              currentAssessmentInitialValues: {
+                assessmentWorkflowId: "some fake id",
+                assessmentDate: addDays(1, new Date()),
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts:530]: Minor, not a blocker, one more if you would like to change to `addDays(1, new Date())`.

File: sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts (hunk 4 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts]: Thanks ❤️

File: sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts (hunk 10 above)
Comment (dheepak-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts]: Modified here and other places of the file.

**Review by andrewsignori-aot — APPROVED** [review]
Thanks for making the changes, looks good 👍

File: sources/packages/backend/apps/queue-consumers/src/processors/assessment/cancel-application-assessment.processor.ts [hunk 11]
```diff
@@ -145,11 +146,19 @@ export class CancelApplicationAssessmentProcessor {
...
       await summary.info("Overawards rollback check executed.");
+      // Check if the assessment has one of the COE(s) declined.
+      const hasDeclinedCOE = assessment.disbursementSchedules.some(
+        (disbursement) => disbursement.coeStatus === COEStatus.declined,
+      );
       if (
         assessment.application.applicationStatus !==
-        ApplicationStatus.Overwritten
+          ApplicationStatus.Overwritten &&
+        !hasDeclinedCOE
```
Comment (sh16011993) [inline sources/packages/backend/apps/queue-consumers/src/processors/assessment/cancel-application-assessment.processor.ts:156]: 👍

File: sources/packages/backend/apps/queue-consumers/test/helpers/testing-modules/testing-modules-helper.ts [hunk 12]
```diff
@@ -44,7 +45,7 @@ export async function createTestingAppModule(): Promise<CreateTestingModuleResul
   );
   const sshClientMock = createMock<Client>();
   const module: TestingModule = await Test.createTestingModule({
-    imports: [QueueConsumersModule],
+    imports: [QueueConsumersModule, DiscoveryModule],
```
Comment (sh16011993) [inline sources/packages/backend/apps/queue-consumers/test/helpers/testing-modules/testing-modules-helper.ts:48]: @dheepak-aot FMU, can you please briefly tell why the `DiscoveryModule` was imported here.

Comment (dheepak-aot) [reply sources/packages/backend/apps/queue-consumers/test/helpers/testing-modules/testing-modules-helper.ts:48]: Discovery module is something that we have been using to get the provider/context from a particular module. 

When we try to get a provider from a nestjs application using app.get(), it does not guarantee if the instance of the provider was returned  from root module, or one of the child modules of the root module.

So we use the discovery service(which require the discovery module) to get a provider instance form a particular module.

File: sources/packages/backend/libs/services/src/confirmation-of-enrollment/confirmation-of-enrollment.service.ts [hunk 13]
```diff
@@ -437,6 +450,23 @@ export class ConfirmationOfEnrollmentService {
...
+            this.systemUserService.systemUser.id,
+            transactionalEntityManager,
+          );
+
+        if (impactedApplication) {
+          this.logger.log(
+            `Application id ${impactedApplication.id} was detected as impacted and will be reassessed.`,
+          );
+        } else {
+          this.logger.log("No impacts were detected on future applications.");
```
Comment (guru-aot) [inline sources/packages/backend/libs/services/src/confirmation-of-enrollment/confirmation-of-enrollment.service.ts:468]: Do we need this else loop just for a log?

**Review by guru-aot — APPROVED** [review]
LGTM, a minor comment not a blocker. Nice work @dheepak-aot

File: sources/packages/backend/libs/services/src/confirmation-of-enrollment/confirmation-of-enrollment.service.ts (hunk 13 above)
Comment (dheepak-aot) [reply sources/packages/backend/libs/services/src/confirmation-of-enrollment/confirmation-of-enrollment.service.ts:468]: Even the if condition here exclusively for logging. I idea here is to be more precise on what happened on assess impacted application.

Please let me know if there is a different suggested approach.

File: sources/packages/backend/apps/queue-consumers/test/helpers/testing-modules/testing-modules-helper.ts (hunk 12 above)
Comment (sh16011993) [reply sources/packages/backend/apps/queue-consumers/test/helpers/testing-modules/testing-modules-helper.ts:48]: Thanks @dheepak-aot . 👍 Nicely explained. 😄

**Review by sh16011993 — APPROVED** [review]
Nice Work @dheepak-aot 👍  . Thanks for clarifying my doubts over the call.
